src/main.rs

```diff
@@ -1,100 +1,170 @@
+    }
+
+    Ok(args)
+}
+
+/// Return `true` if string reference indicates a string with an odd number of characters.
+fn has_odd_character_count(string_ref: &str) -> bool {
+    string_ref.len() % 2 == 1
+}
+
+/// Generate a private key corresponding to a vanity prefix, while search is ongoing.
+///
+/// Once a match is found, a match message is transmitted to the main thread. Once the main thread
+/// has received sufficient match transactions, it broadcasts an exit transaction to search threads.
+///
+/// # Arguments
+///
+/// * `prefix` - The vanity prefix to search against
+/// * `multisig` - If `true` search for a multisig address
+/// * `match_tx` - Transmit channel for match message sent to main thread when a match is found
+/// * `exit_rx` - Receive channel for exit message broadcast from main thread when search is done
+fn generate_key(
+    prefix: String,
+    multisig: bool,
+    match_tx: channel::Sender<()>,
+    exit_rx: channel::Receiver<()>,
+) -> Result<()> {
+    // Clone prefix for getting bytes to check against.
+    let mut check_str = prefix.clone();
+    // Pop last character of check string if an odd length.
+    if has_odd_character_count(&check_str) {
+        check_str.pop();
     };
-    if hex::decode(check_str).is_err() {
-        println!("Vanity prefix must contain valid hex characters (0123456789abcdef).");
-        return; // Exit if vanity prefix isn't specified in valid hex characters.
+
+    // Convert resulting string to bytes for checking against. For odd-length prefixes this approach
+    // truncates the final character, which must be checked via string compare after a bytes match.
+    // This is done for performance (to minimize string operations during the loop).
+    let prefix_bytes = hex::decode(check_str)?;
+
+    // Randomly generate private keys in a loop and check match against prefix bytes.
+    let mut key_generator = KeyGen::from_os_rng();
+    loop {
+        // Exit search loop if exit message has been broadcast.
+        if exit_rx.try_recv().is_ok() {
+            return Ok(());
+        }
+        // Generate a private key and from it, bytes to compare against prefix bytes.
+        let private_key = key_generator.generate_ed25519_private_key();
+        let account_address_bytes = auth_key_bytes_vec(&private_key);
+        let search_bytes = if multisig {
+            create_multisig_account_address(
+                AccountAddress::from_bytes(account_address_bytes)?,
+                SEQUENCE_NUMBER_MULTISIG,
+            )
+            .to_vec()
+        } else {
+            account_address_bytes
+        };
+
+        // If a match found, print diagnostics then keep searching.
+        if search_bytes.starts_with(&prefix_bytes) {
+            let search_str = hex::encode(search_bytes);
+            // Verify full string match only if prefix has an odd character count.
+            if has_odd_character_count(&prefix) && !search_str.starts_with(&prefix) {
```

Comment:
This branch will never run

Reply:
Counterexample:

1. User specifies prefix `fff`
2. `check_str` is thus `ff`
3. `prefix_bytes` is thus `[255]`
4. The address `ffabc...` is mined, which triggers the `search_bytes.starts_with...` check because the first byte is `255` (first `if`)
5. The prefix has an odd character count (3) && the search string (`ffabc...`) does *not* start with `fff` (second `if`)
6. The loop continues to the next iteration